src/plugins/ArithmeticExceptions.cpp

```diff
@@ -0,0 +1,366 @@
+            // Other than that it is not possible to differentiate between signed/unsigned values
+            if(BinOp->hasNoSignedWrap())
+            {
+                const TypedValue LHS = workItem->getOperand(instruction->getOperand(0));
+                const TypedValue RHS = workItem->getOperand(instruction->getOperand(1));
+                const unsigned int ResultWidth = BinOp->getType()->getScalarSizeInBits();
+
+                for(unsigned int i = 0; i < result.num; ++i)
+                {
+
+                    if((RHS.getSInt(i) > 0 && LHS.getSInt(i) < getSignedMinValue(ResultWidth) + RHS.getSInt(i)) ||
+                       (RHS.getSInt(i) < 0 && LHS.getSInt(i) > getSignedMaxValue(ResultWidth) + RHS.getSInt(i)))
+                    {
+                        logArithmeticException();
+                    }
+                }
+            }
+
+            break;
+        }
+        case llvm::Instruction::UDiv:
+        {
+            // Check for division by zero
+            const TypedValue RHS = workItem->getOperand(instruction->getOperand(1));
+
+            for(unsigned int i = 0; i < result.num; ++i)
+            {
+                if(RHS.getSInt(i) == 0)
+                {
+                    logArithmeticException();
+                }
+            }
+
+            break;
+        }
+        case llvm::Instruction::UIToFP:
+        {
+            const auto&& CastInst = llvm::dyn_cast<llvm::CastInst>(instruction);
+            const TypedValue Val = workItem->getOperand(instruction->getOperand(0));
+
+            double MaxValue = 0;
+
+            if(CastInst->getDestTy()->isDoubleTy())
+            {
+                MaxValue = DBL_MAX;
+            }
+            else if(CastInst->getDestTy()->isFloatTy())
+            {
+                MaxValue = FLT_MAX;
+            }
+            else if(CastInst->getDestTy()->isHalfTy())
+            {
+                MaxValue = HALF_MAX;
+            }
+            else
+            {
+                FATAL_ERROR("Unknown float type");
+            }
+
+            if(Val.getUInt(0) > MaxValue)
+            {
+                logArithmeticException();
+            }
+
+            break;
+        }
+        case llvm::Instruction::URem:
+        {
+            // Check for division by zero
+            const TypedValue RHS = workItem->getOperand(instruction->getOperand(1));
+
+            for(unsigned int i = 0; i < result.num; ++i)
+            {
+                if(RHS.getSInt(i) == 0)
+                {
+                    logArithmeticException();
+                }
+            }
+
+            break;
+        }
+        case llvm::Instruction::Unreachable:
+            FATAL_ERROR("Encountered unreachable instruction");
+    }
+}
+
+void ArithmeticExceptions::logArithmeticException() const
+{
+  Context::Message msg(WARNING, m_context);
+  msg << "Undefined behaviour due to an arithmetic exception" << endl
```

Comment:
Can we make this a bit more specific to the particular exception that we are logging? I think it be useful for developers to be able to see the difference between, for example, a division by zero and a signed overflow.